src/pointer-analysis/value_set_domain.h

```diff
@@ -12,64 +12,125 @@ Author: Daniel Kroening, kroening@kroening.com
 #ifndef CPROVER_POINTER_ANALYSIS_VALUE_SET_DOMAIN_H
 #define CPROVER_POINTER_ANALYSIS_VALUE_SET_DOMAIN_H
 
-#define USE_DEPRECATED_STATIC_ANALYSIS_H
-#include <analyses/static_analysis.h>
+#include <analyses/ai.h>
+
+#include <util/invariant.h>
 
 #include "value_set.h"
 
 /// This is the domain for a value set analysis. It is intended to be the
 /// template parameter for `value_set_analysis_templatet`, so `VST` would be
 /// `value_sett`.
-template<class VST>
-class value_set_domain_templatet:public domain_baset
+template <class VST>
+class value_set_domain_templatet : public ai_domain_baset
 {
+protected:
+  /// ait checks whether domains are bottom to increase speed and accuracy.
+  /// Older frameworks don't so it is necessary to track this.
+  bool reachable;
+
 public:
   VST value_set;
 
-  // overloading
+  explicit value_set_domain_templatet(locationt l) : reachable(false)
+  {
+    value_set.clear();
+    value_set.location_number = l->location_number;
+  }
 
-  bool merge(const value_set_domain_templatet<VST> &other, locationt)
+  void make_bottom() override
   {
-    return value_set.make_union(other.value_set);
+    reachable = false;
+    value_set.clear(); //???
   }
 
-  void output(const namespacet &, std::ostream &out) const override
+  void make_top() override
   {
-    value_set.output(out);
+    UNREACHABLE;
   }
 
-  void initialize(const namespacet &, locationt l) override
+  void make_entry() override
   {
-    value_set.clear();
-    value_set.location_number=l->location_number;
+    reachable = true;
+  }
+
+  bool is_bottom() const override
+  {
+    return reachable == false && value_set.values.size() == 0;
+  }
+
+  bool is_top() const override
+  {
+    return false;
+  }
+
+  // overloading
+
+  bool
+  merge(const value_set_domain_templatet<VST> &other, trace_ptrt, trace_ptrt)
+  {
+    reachable |= other.reachable;
+    return value_set.make_union(other.value_set);
+  }
+
+  void
+  output(std::ostream &out, const ai_baset &, const namespacet &) const override
+  {
+    value_set.output(out);
   }
 
   void transform(
-    const namespacet &ns,
     const irep_idt &function_from,
-    locationt from_l,
+    trace_ptrt from,
     const irep_idt &function_to,
-    locationt to_l) override;
+    trace_ptrt to,
+    ai_baset &ai,
+    const namespacet &ns) override;
 
   void get_reference_set(
     const namespacet &ns,
     const exprt &expr,
-    value_setst::valuest &dest) override
+    value_setst::valuest &dest)
   {
     value_set.get_reference_set(expr, dest, ns);
   }
+
+  exprt get_return_lhs(locationt to) const
+  {
+    // get predecessor of "to"
+    to--;
```

Comment:
I am wondering whether `get_return_lhs` should have a different interface: decrementing `to` is an unchecked operation, and would be undefined when `to` already is the beginning of a container.

Reply:
Yes; it is a risky bit of code and far from an ideal design.  It is moved over because I was trying to preserve functionality and structure where I could.  I should refactor this though.

Reply:
I have had a look at this and when doing a "return" edge from `END_FUNCTION`, decrementing `to` really is the best way of getting the function call instruction.  That is probably a wider API failure.  However the only case in which this is actually needed is to handle the return value.  So if we are happy return removal has always been done first then it is fine to just remove all of this code.

Reply:
I am mildly concerned that some of the more obscure users might not ensure this so I think I will just simplify but preserve the functionality.